### ISSUE TITLE
current variableBounds insufficent when using negative lookaheads

### ISSUE THREAD

**Issue by matter123** [jeff-hykin/cpp-textmate-grammar#137]
currently, the bounds are:
- a negative look behind for a word character
- a negative look ahead for a word character

see https://regex101.com/r/F6Cuov/3

the fixed bounds https://regex101.com/r/lQ3Dfr/1

A positive lookahead must be added to the entry bound and the exit bound must be moved inside the lookahead.

**Comment by jeff-hykin** [comment]
Is this causing an issue on any existing pattens? The variableBounds[] wasn't really made to do negative lookaheads on.

**Comment by matter123** [comment]
qualified_type uses the work around.

**Comment by jeff-hykin** [comment]
What do you think about making this a language feature?
```
newPattern( /misc preceding stuff/ ).then(
   match: /identifier/,
   word_cannot_be_any_of: [ "if", "else", "class", "struct" ],
)
```
Which would compile to:
```
/misc preceding stuff\b(?!(?:if|else|class|struct)\b)identifier/
```

**Comment by matter123** [comment]
Looks promising.

**Comment by matter123** [comment]
Shouldn't the first word boundary be inside of the negative look ahead?

**Comment by jeff-hykin** [comment]
Yeah that would make more sense and be more literally true

**Comment by matter123** [comment]
I don't know if the specific example I posted is special but performing the suggested alterations fails to prevent a match of `const`

**Comment by jeff-hykin** [comment]
I think the normal word just needs bounds, this works:
```
(?!\b(static|const)\b)\b\w+\b bar
```

**Comment by jeff-hykin** [comment]
This also works, although I'm not sure if it has better or worse performance
```
\w+(?<!\bstatic\b|\bconst\b) bar
```

**Comment by matter123** [comment]
Are you sure? https://rubular.com/r/vTIIJpQGQVtTIe

**Comment by jeff-hykin** [comment]
Yeah, there shouldn't be any matches since its saying "the word before bar cannot be 'static' or 'const'"

**Comment by matter123** [comment]
regex101 says the lookahead fails in 82 steps while the look behind took 239 steps. Additionally the lookahead matches in 78 while the lookbehind matches in 221.

**Comment by jeff-hykin** [comment]
Yeah I looked at the ms for positive and negative matches and `\w+(?<!\bstatic\b|\bconst\b) bar
` takes ~32ms  and `(?!\b(static|const)\b)\b\w+\b bar` takes ~15ms. So there's a pretty clear winner

**Comment by matter123** [comment]
I'm thinking `:word_cannot_be_any_of` should detect the left word boundary and only apply a left word boundary to the lookahead if there is one on the match. I don't yet have a use case but it should not break the existing use cases.

**Comment by jeff-hykin** [comment]
implemented with `Fix/#316`
